src/serac/physics/state/finite_element_vector.cpp

```diff
@@ -150,4 +150,19 @@ double min(const FiniteElementVector& fe_vector)
   return global_min;
 }
 
+double innerProduct(const FiniteElementVector& v1, const FiniteElementVector& v2)
+{
+  SLIC_ERROR_IF(
+      v1.Size() != v2.Size(),
+      axom::fmt::format("Finite element vector of size '{}' can not inner product with another vector of size '{}'",
+                        v1.Size(), v2.Size()));
+  SLIC_ERROR_IF(v1.comm() != v2.comm(),
+                "Can not compute inner products between vectors with different mpi communicators");
+
+  double global_ip;
+  double local_ip = mfem::InnerProduct(v1, v2);
```

Comment:
Is there a reason this doesn't work? https://github.com/mfem/mfem/blob/1ccb31fde641a96c9fd2920bc123c35eee31d476/linalg/hypre.hpp#L333 Is it resolving to the non-reduced `mfem::Vector` implementation if we don't force it?

Reply:
It didn't call that one by default, it was calling the non-reduced one for sure.  Honestly, I didn't know that one existed!  Not always clear which vector in mfem has what implemented for it.  I think there is likely a way to force it to call that one (now that I know it exists).  But, this exercise has given me a new mission for serac!  I would love to convince everyone that our FiniteElementVector should have a HypreParVector, not be one.  There are many reasons for this, e.g., good dependency management through interfaces, ..., but mostly it allows us to control exactly how common math operations work in a consistent way, without having to always think about which vector versions is being invoked in the background depending on the input types.  If we do not agree with certain mfem conventions (like not overloading +/-), we can simply do that for our interface layer independently of any current or changing mfem decisions on such things.

Reply:
I think the biggest blocker for this is we would have to call the accessor for the underlying `mfem::HypreParVector` to interact with `Functional` or any of the nonlinear, linear, or ODE solvers. That said, it would make reasoning about these math operations much simpler for the reasons you described. I'll tag this as a discussion topic for Friday.